src/System.Private.CoreLib/shared/System/Number.NumberToDouble.cs

```diff
@@ -0,0 +1,450 @@
+            }
+
+            Debug.Assert(!isBad, "NumberToDouble table not correct. Correct version dumped to debug output.");
+        }
+#endif
+
+        // get 32-bit integer from at most 9 digits
+        private static uint DigitsToInt(char* p, int count)
+        {
+            Debug.Assert((1 <= count) && (count <= 9));
+
+            char* end = (p + count);
+            uint res = (uint)(p[0] - '0');
+
+            for (p++; p < end; p++)
+            {
+                res = (10 * res) + p[0] - '0';
+            }
+
+            return res;
+        }
+
+        private static int GetLength(char* src)
+        {
+            int length = 0;
+
+            while (src[length] != '\0')
+            {
+                length++;
+            }
+
+            return length;
+        }
+
+        // helper function to multiply two 32-bit uints
+        private static ulong Mul32x32To64(uint a, uint b)
+        {
+            return (ulong)(a) * b;
+        }
+
+        // multiply two numbers in the internal integer representation
+        private static ulong Mul64Lossy(ulong a, ulong b, ref int exp)
+        {
+            // it's ok to lose some precision here - Mul64 will be called
+            // at most twice during the conversion, so the error won't propagate
+            // to any of the 53 significant bits of the result
+            ulong val = Mul32x32To64((uint)(a >> 32), (uint)(b >> 32))
+                      + (Mul32x32To64((uint)(a >> 32), (uint)(b)) >> 32)
+                      + (Mul32x32To64((uint)(a), (uint)(b >> 32)) >> 32);
+
+            // normalize
+            if ((val & 0x8000000000000000) == 0)
+            {
+                val <<= 1;
+                exp--;
+            }
+
+            return val;
+        }
+
+        // slower high precision version of Mul64 for computation of the tables
+        private static ulong Mul64Precise(ulong a, ulong b, ref int exp)
+        {
+            ulong hilo = ((Mul32x32To64((uint)(a >> 32), (uint)(b)) >> 1)
+                       + (Mul32x32To64((uint)(a), (uint)(b >> 32)) >> 1)
+                       + (Mul32x32To64((uint)(a), (uint)(b)) >> 33)) >> 30;
+
+            ulong val = Mul32x32To64((uint)(a >> 32), (uint)(b >> 32))
+                      + (hilo >> 1)
+                      + (hilo & 1);
+
+            // normalize
+            if ((val & 0x8000000000000000) == 0)
+            {
+                val <<= 1;
+                exp--;
+            }
+
+            return val;
+        }
+
+        private static double NumberToDouble(ref NumberBuffer number)
+        {
+#if DEBUG
+            
+            if (!s_CheckedTables)
+            {
+                CheckTables();
+                s_CheckedTables = true;
+            }
+#endif
+
+            char* src = number.digits;
+            int total = GetLength(src);
+            int remaining = total;
+
+            // skip the leading zeros
+            while (src[0] == '0')
```

Comment:
Are we guaranteed that the string at least 1 character long?

Reply:
Same comment as https://github.com/dotnet/coreclr/pull/20080#discussion_r219332505